newrelic/hooks/mlmodel_openai.py

```diff
@@ -0,0 +1,145 @@
+    span_id = available_metadata.get("span.id", "")
+    trace_id = available_metadata.get("trace.id", "")
+
+    response_headers = getattr(response, "_nr_response_headers", None)
+    response_model = response.model
+    response_id = response.get("id", "")
+    request_id = response_headers.get("x-request-id", "")
+    settings = transaction.settings if transaction.settings is not None else global_settings()
+
+    chat_completion_summary_dict = {
+        "id": chat_completion_id,
+        "appName": settings.app_name,
+        "conversation_id": conversation_id,
+        "span_id": span_id,
+        "trace_id": trace_id,
+        "transaction_id": transaction._transaction_id,
+        "request_id": request_id,
+        "api_key_last_four_digits": f"sk-{response.api_key[-4:]}",
+        "duration": ft.duration,
+        "request.model": kwargs.get("model") or kwargs.get("engine"),
+        "response.model": response_model,
+        "response.organization":  response.organization,
+        "response.usage.completion_tokens": response.usage.completion_tokens,
+        "response.usage.total_tokens": response.usage.total_tokens,
+        "response.usage.prompt_tokens": response.usage.prompt_tokens,
+        "request.temperature": kwargs.get("temperature", ""),
+        "request.max_tokens": kwargs.get("max_tokens", ""),
+        "response.choices.finish_reason": response.choices[0].finish_reason,
+        "response.api_type": response.api_type,
+        "response.headers.llmVersion": response_headers.get("openai-version", ""),
+        "response.headers.ratelimitLimitRequests": check_rate_limit_header(response_headers, "x-ratelimit-limit-requests", True),
+        "response.headers.ratelimitLimitTokens": check_rate_limit_header(response_headers, "x-ratelimit-limit-tokens", True),
+        "response.headers.ratelimitResetTokens": check_rate_limit_header(response_headers, "x-ratelimit-reset-tokens", False),
+        "response.headers.ratelimitResetRequests": check_rate_limit_header(response_headers, "x-ratelimit-reset-requests", False),
+        "response.headers.ratelimitRemainingTokens": check_rate_limit_header(response_headers, "x-ratelimit-remaining-tokens", True),
+        "response.headers.ratelimitRemainingRequests": check_rate_limit_header(response_headers, "x-ratelimit-remaining-requests", True),
+        "vendor": "openAI",
+        "ingest_source": "Python",
+        "number_of_messages": len(kwargs.get("messages", [])) + len(response.choices),
+    }
+
+    transaction.record_ml_event("LlmChatCompletionSummary", chat_completion_summary_dict)
+    message_list = list(kwargs.get("messages", [])) + [response.choices[0].message]
+
+    create_chat_completion_message_event(transaction, settings.app_name, message_list, chat_completion_id, span_id, trace_id, response_model, response_id, request_id)
+
+    return response
+
+
+def check_rate_limit_header(response_headers, header_name, is_int):
+    if not response_headers:
+        return ""
```

Comment:
If we return an empty string, this attribute is not stored on the event. This seems fine to me, just confirming we are in agreement that that's the behavior we want.

Reply:
Checking with the AI team on this

Reply:
This behavior is confirmed to be fine 👍🏽